### DISCUSSION TITLE
CMake version of Fpm

### DISCUSSION THREAD

**Discussion by ivan-pi** [fortran-lang/fpm#1080] [General]
If anyone would like to experiment with using CMake to build fpm, I created a version at my branch here: https://github.com/ivan-pi/fpm/tree/cmake

The dependencies have to be cloned manually with the commands:
```
$ git clone https://github.com/ivan-pi/fpm.git && cd fpm
$ git checkout cmake && cd dep
$ source clone_deps.sh
$ cd ..
```
Afterward you can build as usual:
```
$ cmake -B build -G Ninja -DCMAKE_Fortran_COMPILER=gfortran -DCMAKE_C_COMPILER=gcc
$ cd build
$ ninja -j8
```

I was using CMake + Ninja + [ninjatracing](https://github.com/nico/ninjatracing) to investigate build dependencies using the [Perfetto](https://ui.perfetto.dev/) trace viewer. The trace shown below includes the dependencies. 

![image](https://github.com/user-attachments/assets/c3b482b1-4b1f-4323-8e50-9d10369cc568)

**Comment by awvwgk** [comment]
Looks like we have some rather big chunks of code which need to be compiled sequentially. Are there any straightforward candidates which could be split in several files or separated as submodules to make compilation faster?

**Reply by ivan-pi** [reply]
In principle everything could be done parallel with decomposition into modules and submodules. But the entire build time is < 20s, and generally less if you are in a edit compile run cycle. So it is not that critical. 

[fpm_trace.json](https://github.com/user-attachments/files/17132406/fpm_trace.json)

I've attached a json file, you just need to import it into [ui.perfetto.dev](https://www.ui.perfetto.dev/) to study the build.

**Reply by ivan-pi** [reply]
The first chokepoint is the direct chain,
- `fpm_strings`: https://github.com/fortran-lang/fpm/blob/main/src/fpm_strings.f90
- `fpm_error`: https://github.com/fortran-lang/fpm/blob/main/src/fpm/error.f90
- `fpm_toml`: https://github.com/fortran-lang/fpm/blob/main/src/fpm/toml.f90

`fpm_toml` depends on `tomlf` and `jonquil`. The dependency on `tomlf` is a logical one, because the `toml_table` is part of the interface. On the other hand the dependency on `jonquil` could be "buried" in the submodule.

The next large chokepoint are the two modules:
- `fpm_manifest_package`: https://github.com/fortran-lang/fpm/blob/main/src/fpm/manifest/package.f90) 
- `fpm_manifest`: https://github.com/fortran-lang/fpm/blob/main/src/fpm/manifest.f90)

which depend on the cluster of modules in [src/fpm/manifest
](https://github.com/fortran-lang/fpm/tree/main/src/fpm/manifest).

If my understanding is correct, if everything along the critical path on Thread 0 was split into modules/submodules, then the modules would be compiled quickly, and the rest of the build would be free to proceed until new modules became the bottlenecks.

**Reply by ivan-pi** [reply]
I've tried splitting `fpm_toml` into module and submodule.

Here is a zoomed-in view before the split:

![image](https://github.com/user-attachments/assets/45a82890-11eb-4fd0-b08b-c1bc3bd74e14)

And here is after the split (note, time-axis is not the same):

![image](https://github.com/user-attachments/assets/cba1f1d8-65f0-4104-bbea-ea28857916d3)

You can see the submodule now depends on the parent module.

So if this were done for the modules that take longer, it would deliver some speed-up.

**Reply by ivan-pi** [reply]
I decided to test this hypothesis now on the module `fpm` which is at the end of the chain. In practice the splitting doesn't help much, because the submodule still has all the same dependencies it had before. So it makes sense to address modules early in the chain.

**Reply by ivan-pi** [reply]
I've tried restructuring `fpm_manifest` to group all the config types together, and then leave the implementation of those classes for submodules. 

The build trace now looks like this:

![image](https://github.com/user-attachments/assets/3a6a6837-128f-4d07-9903-25a52e842f20)

vs without submodules:

![image](https://github.com/user-attachments/assets/304f4137-79bf-4c96-b66f-94ef276eb9b0)

Build time is down from 12s to 3s.

**Reply by ivan-pi** [reply]
The submodule version can be found here: https://github.com/ivan-pi/fpm/tree/sm

The build trace file: [fpm_more_sm.json](https://github.com/user-attachments/files/17167783/fpm_more_sm.json)

To my great delight, the CMake + submodules version can still build itself the "fpm" way.
